enhancements/installer/component-selection.md

```diff
@@ -0,0 +1,500 @@
+  exclude:
+  - console
+  - samples
+  include:
+  - registry
+```
+
+Which resources ultimately get installed for a given cluster would be the set of resources encompassed
+by the CLUSTER_PROFILE(if any), minus any resources explicitly excluded by the excluded capabilities configuration,
+plus any explicitly included capabilities.  Any resource that is not excluded by the CLUSTER_PROFILE, is not
+included via the Include stanza, and is not Excluded via the Exclude stanza, will be filtered or not based on
+the `inclusionDefault` value (Include or Exclude).  The `inclusionDefault` field itself will default to `Include`.
+
+The `inclusionDefault` field (carried through to the CVO) allows an Admin to choose what behavior they want for new
+capabilities that might come as part of a future upgrade (either have the new capability installed during
+the upgrade, or exclude it)
+
+Specifying the same capability as an `include` and an `exclude` is an error condition.
+
+
+Examples of candidate components to be treated as capabilities:
+
+* console
+* imageregistry
+* samples
+* cluster baremetal operator
+* olm/marketplace
+* csi-*
+* insights
+* monitoring
+* ???
+
+
+
+3) Pass the list of filtered annotations to the CVO.  This is distinct from overrides because overrides
+put the cluster in an unsupported state.  Filtered annotations are supported for upgrades.  The filtered
+components will be listed in the [ClusterVersion](https://github.com/openshift/api/blob/4436dc8be01e8dcd8b250e1b32bb0fbd64ba78ac/config/v1/types_cluster_version.go#L35) object:
+
+```yaml
+  spec:
+    capabilities:
+      inclusionDefault: [Include|Exclude]
+      exclude:
+      - console
+      - samples
+      include:
+      - registry
+```
+
+The CVO will filter out(not apply/reconcile) resources that are annotated with key/value
+`capability.openshift.io/name=$exclusions` and listed in the `exclude` section, or are not listed
+and the `default` setting is `exclude`.  In addition any explicitly included capability
+should be added, regardless of the CLUSTER_PROFILE.
+
+If a resource has multiple `capability.openshift.io/name` annotations, then the resource will be filtered if any
+of the annotations on the resource are matched by a configured filter.  This allows a component to define
+a `ServiceMonitor` which is part of both the `Monitoring` capability and the `Foo` capability.  If either
+of those capabilities are excluded, the `ServiceMonitor` should not be created.  (If Monitoring is excluded
+then no ServiceMonitors should be created, if only Foo is excluded, then only Foo's ServiceMonitors should
+be excluded).
+
+Specifying the same capability as an `include` and an `exclude` is an error condition.  The CVO will need to
+implement a validating webhook on ClusterVersion to enforce this.  This avoids the need to define behavior for
+what to do when a user specifies the same capability in both, and update a status field indicating what choice
+the CVO made.
```

Comment:
We've managed to avoid growing a validating webhook so far.  Some possible alternatives:

a. If a user sets conflicting values, we can continue with whatever the previous setting was in memory (until the next CVO container restart) and complain about the broken config in `status.conditions[Type=Failing]`, hoping the user notices and fixes the config before the next CVO-container restart.
b. Shuffle the API around a bit to make it impossible to add broken data.  For example, you could have:

```yaml
spec:
  capabilities:
  - console: [Include|Exclude]
  - ...
  capabilitiesInclusionDefault: [Include|Exclude]
```

And with the API-server enforcing that `Include`/`Exclude` enum, folks would be unable to push broken data.  I think; I'm not actually sure the API-server rejects maps/objects with repeat keys like:

```yaml
spec:
  capabilities:
  - console: Include
  - console: Exclude
```

Thoughts?

Reply:
aren't you going to need a validating webhook anyway to prevent someone from changing a value from "include" to exclude", or generally adding a new exclusion?  (see a few paragraphs down)

Reply:
I don't think so, because you could have the CVO just reject the misconfigured `spec` and complain in `conditions` or whatever, which we'll need anyway, in case someone gets some bad data around the webhook

Reply:
will update this to say we'll allow anything in spec, and only copy valid spec to status, cvo will be driven off status, and status will contain an error indicator.